src/app/shared/services/data/app-data-variable.service.ts

```diff
@@ -120,7 +121,10 @@ export class AppDataVariableService extends AsyncServiceBase {
           isRecursive = true;
         }
         for (const variableName of Object.keys(variableNameHashmap)) {
-          const evaluated = await this.get(contextPrefix as IVariableContext, variableName);
+          let evaluated = await this.get(contextPrefix as IVariableContext, variableName);
+          if (typeof evaluated === "string" || evaluated instanceof String) {
+            evaluated = `"${evaluated}"`;
```

Comment:
nit(blocking) - This is quite a significant change to a core method. Could you try to identify the use case where this is required and add to the existing spec tests for the service?

You'll also see that this change breaks the existing tests by adding extra quotation marks where not required.

I think this highlights an importance to try get the ng test suite fully operational (some work started in #2080 which was meant to be handed over, but possibly wasn't or fell down priority list?)

Reply:
I've added comments to the app-data and app-data-variable spec files to include code to run test standalone (e.g. `yarn ng test --include src\app\shared\services\data\app-data-variable.service.spec.ts`

Reply:
Sure enough, I had misunderstood the issue. 

I was getting the following log when trying to run the example_temp_override_default template:
<img width="500" alt="Screenshot 2023-11-07 at 16 50 01" src="https://github.com/IDEMSInternational/parenting-app-ui/assets/64838927/f95fc9a1-8ab6-4885-903d-51bc696514ae">
The override condition, `@fields._app_language.startsWith("es")` (found in the [contents list](https://docs.google.com/spreadsheets/d/1MpoH3BxhECZRmYM10HZ0pTOoe69FJ-fEW9FwzK-Q6yw/edit#gid=1745157248)) was being parsed as `za_en.startsWith("es")`, which I took to be a general issue with fields storing string values. In fact, I think it is general behaviour affecting all cases of _applying a string method_ to the value of a field. The issue can be resolved by rewriting the condition as `"@fields._app_language".startsWith("es")`. Do you think this is expected behaviour? That debug template appears to be the only template that uses `startsWith()`.

I have reverted the change to this method in the latest commit

Reply:
Interesting, does it work if using delimited syntax `{@fields._app_language}.startsWith("es")`?
Not super sure what the expected behaviour here is, might be good to write a couple tests and update template as required

Reply:
I've added some tests to [src/app/shared/services/data/app-data-variable.service.spec.ts](https://github.com/IDEMSInternational/parenting-app-ui/pull/2107/files#diff-8873dfedef4ee3fd95f8af1a3d11a9856aff28b2c9cdc6683952d0daa53493cc), which capture the current behaviour.

<img width="800" alt="Screenshot 2023-11-09 at 15 15 57" src="https://github.com/IDEMSInternational/parenting-app-ui/assets/64838927/36293753-32bd-48f7-8fb9-ceb3fe8fd74c">

I think this is not expected behaviour, and we should amend the app data variable service, see new issue: #2141

Reply:
Me and Esmee are wondering if it might make sense for condition fields to require an `@calc()` block in order to pass javascript. As I understand it, our `condition` columns (`override_condition` and `condition_list` on campaigns) evaluate any javascript that is passed. But it could be confusing for authors to understand when they can and can't write javascript directly in a cell

Reply:
Ah, yeah I see the challenge here - comparing the two cases:

1. `hello @field.name`
2. `@field.name.startsWith("A")`

In case (1) we want the string to just replace and return `hello my_name`
In case (2) we want the string to replace wrapped as a string `"my_name".startsWith("A")`

Currently we focus on supporting case (1). The parser attempts to evaluate, but reads `hello` and `my_name` as variables which are not defined, and has a fallback in case of such errors to just return the full expression as-is (as text)

I think the best way to also support case (2) would be to internally store a mapping of replaced strings when evaluating, which can then be used in follow-ups. E.g. for both cases above the mapping would simply be variables
```
{my_name: "my_name"}
```
Then when it comes to evaluation, case 1 will still just return as text (failing to recognise `hello` as a variable), but case 2 should be able to parse `my_name.startsWith("A")`, substituting the `my_name` variable for `"my_name"` string

I'll try make these changes in a follow-up PR as not so much targetted on the issue at hand

Reply:
Moving over to https://github.com/IDEMSInternational/parenting-app-ui/pull/2146